### DISCUSSION TITLE
Ship sprite scaling issues?

### DISCUSSION THREAD

**Discussion by esksp** [endless-sky/endless-sky#10713] [Q&A]
I don't yet have an eye for ship sizes, so I rendered them a little bigger than needed.

For one ship **I successfully used the `"scale" 0.75` tag** after the sprite declaration. **Works just fine** in the game.

For another ship, the exact same tag (copy-paste) yielded a strange result: **The ship sprite is smaller, as expected, but the ship's outline and the turret placements are not scaled down!**

How comes the same work flow yields two different results?

Here are the ships' codes (just the beginning):
```
ship Nova
	sprite "ship/nova"
		"scale" 0.75
	thumbnail "thumbnail/nova"
	attributes
		category "Heavy Warship"
```
```
ship Supernova
	sprite "ship/supernova"
		"scale" 0.75
	thumbnail "thumbnail/supernova"
	attributes
		category "Superheavy"

```
Is there a rational explanation?

**Comment by TheGiraffe3** [comment]
Outline?
It makes sense that the turrets don't scale, because they're hardcoded.

**Reply by esksp** [reply]
> Outline?

The green flagship outline.

> It makes sense that the turrets don't scale, because they're hardcoded.

Maybe, but what doesn't make any sense is that I have two ships which display a different behavior!...
I would understand the second case, since the turrets have hardcoded coordinates, but when it worked for the first ship I assumed the game engine also scales those coordinates down (fairly simple math after all). Except it didn't for the second ship.

Besides, what's the point of the sprite sizing if turrets and outlines don't follow?

---Edit:
Uh-oh, I realized there is a misunderstanding about the turret scaling: I don't mean the **turret sprites** scaling, but the turret **placements**.

**Comment by warp-core** [comment]
It would be useful to see the full definitions of the ships (or at least, include the hardpoint definitions) as well as screenshots of the ships and hardpoints sprites (showing the hardpoint locations) and the ship info panels.
It would also be useful to see the same hardpoint definitions applied to other ship definitions with unscaled sprites.
Screenshots of the outline behaving incorrectly would also be useful (although, I'm fairly confident I know what's going on there already).

Regarding the hardpoint positions, as I recall, the coordinates are not supposed to be affected by the scale of the sprite.
That is, if you place a hardpoint at (0, -100), it will be 100 pixels "up" from the centre of the sprite, regardless of the scale of the sprite. You are scaling the sprite, not the entire ship definition.

**Reply by warp-core** [reply]
https://github.com/endless-sky/endless-sky/pull/10715
Should fix the outline issue.

**Reply by esksp** [reply]
> It would also be useful to see the same hardpoint definitions applied to other ship definitions with unscaled sprites. 

When not scaled down, everything is where it should be. The placements just don't scale, ~despite doing so for its siblings.~


> Regarding the hardpoint positions, as I recall, the coordinates are not supposed to be affected by the scale of the sprite. That is, if you place a hardpoint at (0, -100), it will be 100 pixels "up" from the centre of the sprite, regardless of the scale of the sprite. You are scaling the sprite, not the entire ship definition.

Really? That radically diminishes the point of scaling, given _all_ ships have hardcoded gun/turret/engine hardpoints. How is this supposed to work?
~And -- most of all -- why does it work on other previous ships???~

Here is the definition of the big one (I removed the outfits to make it shorter):

```
ship Supernova
	sprite "ship/supernova"
		"scale" 0.75
	thumbnail "thumbnail/supernova"
	"uncapturable"
	attributes
		category "Superheavy"
		"cost" 100000000
		licenses
			"Strike Forces"
		"shields" 32000
		"hull" 9800
		"required crew" 55
		"bunks" 255
		"mass" 1240
		"drag" 3
		"heat dissipation" .5
		"fuel capacity" 1000
		"cargo space" 50
		"outfit space" 2800
		"weapon capacity" 1200
		"engine capacity" 990
		"ion resistance" 0.1
		"scramble resistance" 0.2
		"slowing resistance" 3
		"spinal mount" 1
		weapon
			"blast radius" 990
			"shield damage" 19930
			"hull damage" 11965
			"hit force" 11495
	outfits
		...

	engine -23 194
	engine -13 194
	engine -4 194
	engine 5 194
	engine 15 194
	engine 24 194
	engine 33 194
	engine 42 194
# left
	gun -3.5 -172 "Torpedo Launcher"
	gun -3 -172 "Strike Forces Particle Cannon"
	gun -1 -174 "Strike Forces Particle Cannon"
	gun -0.5 -174 "Strike Forces Particle Cannon"
	gun 5.5 -176 "Strike Forces Particle Cannon"
	gun 5 -176 "Strike Forces Particle Cannon"
# right
	gun 6.5 -172 "Strike Forces Particle Cannon"
	gun 6 -172 "Strike Forces Particle Cannon"
	gun 8 -174 "Strike Forces Particle Cannon"
	gun 8.5 -174 "Strike Forces Particle Cannon"
	gun 14.5 -176 "Strike Forces Particle Cannon"
	gun 14 -176 "Torpedo Launcher"
# turrets
	turret -11 -149 "Strike Forces Turret"
	turret 9 -150.5 "Strike Forces Turret"
	turret 30 -149 "Strike Forces Turret"
	turret 9 -82 "Strike Forces Turret"
	turret -17 -37 "Heavy Anti-Missile Turret"
	turret 36 -37 "Heavy Anti-Missile Turret"
	turret 9 19 "Strike Forces Turret"
	turret -29 80.5 "Heavy Anti-Missile Turret"
	turret 47.5 80.5 "Heavy Anti-Missile Turret"
	turret 9 175.5 "Strike Forces Turret"
	turret -103 100 "Strike Forces Turret"
	turret 122 100 "Strike Forces Turret"
# big turret
	turret 9 81.5 "Planetary Destruction Gun"

	explode "small explosion" 25
	explode "medium explosion" 35
	explode "large explosion" 45
	explode "huge explosion" 30
	"final explode" "final explosion large"
	description "Strike Forces Superheavy Cruiser, sporting the absolutely deadly 'Planetary Destruction Gun'. There were only 3 Supernova class ships built, one was destroyed."
	description "You want one? Are you kidding? This government only stuff."
 
```
There is an offset to the coordinates, for some stupid reason I didn't render it centered, so the middle of the ship is at "9". Doesn't matter, since it's just a placeholder sprite (will greeble it more extensively to better show the sheer size). Right now I'm still checking how it handles (thus the generous allocation on various spaces). 

(For the record, it's an unique non-player-only ship patrolling the Strike Forces System (fleet), actively discouraging the player from doing something stupid (as you see it won't go down easily, even in the late game. The "Strike Forces Particle gun" and assorted turrets are lighter versions of the particle gun. The "Planetary Destruction Gun" is a big bad uncapturable spinal mount turret able to destroy most ships with one shot. Think Archon gun.)

> #10715
> Should fix the outline issue.

What's a "Pull Request"? Is this a known bug? Sorry, out of my depth here... :-D
~Whatever it might be, why don't my other two scaled ships (all at 0.75) have the same problem then?~

--- Edit:
**Actually my other ships are not scaled.** They are scaled in the definitions, but when I check my test savegames, they aren't. I must have edited the files at some point but forgot to edit them all...

Anyway, I just checked in-game, and indeed, **they all display the scaling issue**: Outline and turret placements are _not_ scaled. 

--- Edit 2:

Here is a pic. Both are scaled to 0.75:
![ScalingIssues](https://github.com/user-attachments/assets/be68787f-0a5a-42fb-a55b-b2826ae0f992)

**Reply by warp-core** [reply]
> > Regarding the hardpoint positions, as I recall, the coordinates are not supposed to be affected by the scale of the sprite. That is, if you place a hardpoint at (0, -100), it will be 100 pixels "up" from the centre of the sprite, regardless of the scale of the sprite. You are scaling the sprite, not the entire ship definition.
> 
> Really? That radically diminishes the point of scaling, given _all_ ships have hardcoded gun/turret/engine hardpoints. How is this supposed to work?

I don't know what "hardcoded" means here. The hardpoint coordinates can be changed at the same time as sprite scale.

> > #10715
> > Should fix the outline issue.
> 
> What's a "Pull Request"? Is this a known bug? Sorry, out of my depth here... :-D ~Whatever it might be, why don't my other two scaled ships (all at 0.75) have the same problem then?~

GitHub can explain better than me: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests

**Reply by esksp** [reply]
> I don't know what "hardcoded" means here. The hardpoint coordinates can be changed at the same time as sprite scale.

You mean manually? Well, a computer would be able to do that _so_ much quicker than me...  :-D
My point is, the sprite rescaling feature makes only sense if it rescales _everything_ for you, picture _and_ coordinates. After all, the sprite picture itself can also be changed by hand, and it's even easier than recalculating the hardpoint coordinates: Just open the sprite image in some picture editing software, resize it by x%, and save it...

I understand the difficulty of finding and parsing a set (and the right one!) of coordinates inside a "ships" list file, so that should probably just happen automatically on display: "x = "scale", if scale ="" then scale =1, multiply all hardpoint coordinates by x" (or something like that).


> GitHub can explain better than me:

Thanks.
Still too specialized for me. I'd rather keep to the discussion and issue boards, I'm not a developer.  :-)
